packages/pigeon/lib/ast.dart

```diff
@@ -82,39 +74,51 @@ class Api extends Node {
   }
 }
 
-/// Represents a field on a [Class].
-class Field extends Node {
-  /// Parametric constructor for [Field].
-  Field({
-    required this.name,
-    required this.dataType,
+/// A specific instance of a type.
+class TypeDeclaration {
+  /// Constructor for [TypeDeclaration].
+  TypeDeclaration({
+    required this.baseName,
     required this.isNullable,
     this.typeArguments,
-    this.offset,
   });
 
-  /// The name of the field.
-  String name;
+  /// The datatype base name of the entity (ex 'Foo' to 'Foo<Bar>?').
+  final String baseName;
 
-  /// The data-type of the field (ex 'String' or 'int').
-  String dataType;
+  /// The type arguments to the entity.
+  final List<TypeDeclaration>? typeArguments;
```

Comment:
I missed this earlier: do null and empty mean different things here? If not, I would highly recommend making this non-nullable (with a default value of empty list in the constructor if not required), and adjusting the helper function that you use to populate it to always return a non-null value.

When we were doing NNBD migrations in plugins, and in discussions with the Dart team about null safety, the general consensus seemed to be that it's usually better to make collections non-nullable, because if you don't specifically want null and empty to be different, then in practice you have two ways of expressing the same conceptual value (here, "no arguments"), and code that wants to handle that may incorrectly handle only one of the two versions. Having a policy of making collections non-nullable unless null and empty were different found and fixed a few latent bugs (of a type we've found the hard way in several cases pre-NNBD), and often made code a little simpler as well.

I think there may actually be some of those bugs in this PR (unless the distinction is meaningful and I'm just not clear on what that distinction is). `_makeGenericCastCall`, for instance, is documented as "Returns an empty string if the type has no type arguments." but it actually only returns an empty string if `typeArguments` is null. If the array is present but empty (which I would also assume means "no type arguments"), it returns `.cast<>()`.

Reply:
Good call.  Let me land this and I'll follow up with this like the no name parameter issue.

Reply:
I did this in a [followup PR](https://github.com/flutter/packages/pull/435).  I also filed a bug against Dart for not helping me with the transition: https://github.com/dart-lang/sdk/issues/46907